### PR TITLE
Worker heartbeat-based cleanup (2nd try)

### PR DESCRIPTION
**Pull request by fw42**
This is another try at https://github.com/resque/resque/pull/1207 (basically cherry-picked https://github.com/Shopify/resque/commit/cc245ca61a9621e2a6f46defdc2095d3f8ca5f3b with a bit of cleaning up).

@sirupsen @dylanahsmith @steveklabnik

### CODE REVIEW

File: lib/resque.rb [hunk 1]
```diff
@@ -152,6 +152,19 @@ def redis_id
...
 
+  DEFAULT_HEARTBEAT_INTERVAL = 60
+  DEFAULT_PRUNE_INTERVAL = DEFAULT_HEARTBEAT_INTERVAL * 5
+
+  attr_writer :heartbeat_interval
+  def heartbeat_interval
+    @heartbeat_interval || DEFAULT_HEARTBEAT_INTERVAL
+  end
+
+  attr_writer :prune_interval
```
Comment (Sirupsen) [inline lib/resque.rb]: Why not make this an accessor?

**Comment by Sirupsen** [comment]
LGTM

**Comment by Sirupsen** [comment]
For context, this was deployed for Shopify two years ago when we started having hostnames that disappear and reappear constantly due to a Docker-based architecture.

File: lib/resque.rb (hunk 1 above)
Comment (fw42) [inline lib/resque.rb]: Ask yourself, you wrote this :-) But the reader has a fallback to the default, so I guess that's why?

Comment (Sirupsen) [inline lib/resque.rb]: Yeah, your argument for my code makes sense :p :+1: Past Simon

**Comment by steveklabnik** [comment]
Thank you so much for these PRs! I'm going to check them out tonight, or maybe tomorrow. Still at work-work right now.

**Comment by Sirupsen** [comment]
@steveklabnik does a Slack room or similar exist for the people who're interested in resurrecting Resque to discuss direction etc?

**Comment by steveklabnik** [comment]
There is #resque on freenode that I idle in. Frankly, it's been basically just me for a long time, and so if people are interested in doing active work, we should dtermine

1. exactly who that is
2. how we, as a group, want to handle it.

**Comment by steveklabnik** [comment]
This has a merge conflict, but I've re-started the one failing test, seems spurious.

**Comment by fw42** [comment]
> This has a merge conflict

Hm, my GitHub UI shows me this:

screen shot 2016-01-15 at 12 42 57

What's the conflict?

**Comment by steveklabnik** [comment]
I ... am now seeing that too. What the heck. Sorry, must have been confused.

Test is still failing though :/

**Comment by steveklabnik** [comment]
@fw42 any chance you'll be able to update this PR sometime soon? I'd like to maybe get a release out in the next week or two.

**Comment by fw42** [comment]
yeah will take a look today

**Comment by steveklabnik** [comment]
:heart: 

For my part, the failing test was Travis being spurious, but it looks like in the meantime, something I merged caused a conflict. So it should be fairly easy.

Is there anything else that would need to be in the next release to get you all off of your fork?

**Comment by fw42** [comment]
Rebased. No more conflicts. Should be good to go now, but looks like Travis timed out again...

**Comment by steveklabnik** [comment]
Weird that it seems to be 1.9.2 all the time.

I'm sure this will end up passing, but am gonna re-start so we can be sure. :smile:

**Comment by steveklabnik** [comment]
Oh no, 1.9.2 seems to be... doing it again. I am now fearing some sort of obscure, terrible bug.

One thing I've been thinking about is Ruby version compatibility. 1.9.2 is EOL'd as of July 31, 2014. Anyone who is using 1.9.2 can probably upgrade to 1.9.3 by now, which is also EOL'd. I've been considering relaxing requirements to the last version of each line, as especially post Ruby 2.0, upgrades have seemingly been a lot smoother.

I think I'm going to merge this, and then consider the versioning question in another issue.

Thanks again! If nothing comes up, I'll probably release a new version as soon as I fix https://github.com/resque/resque/pull/1206, probably next week at the earliest.

**Pull request merged**